File: di-ipv-cimit-stub/lambdas/put-contra-indicators/src/main/java/uk/gov/di/ipv/core/putcontraindicators/PutContraIndicatorsHandler.java
package uk.gov.di.ipv.core.putcontraindicators;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.nimbusds.oauth2.sdk.util.StringUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.message.StringMapMessage;
import uk.gov.di.ipv.core.library.exceptions.FailedToParseRequestException;
import uk.gov.di.ipv.core.putcontraindicators.domain.PutContraIndicatorsRequest;
import uk.gov.di.ipv.core.putcontraindicators.domain.PutContraIndicatorsRequestBody;
import uk.gov.di.ipv.core.putcontraindicators.domain.PutContraIndicatorsResponse;
import uk.gov.di.ipv.core.putcontraindicators.exceptions.CiPutException;
import uk.gov.di.ipv.core.putcontraindicators.service.ContraIndicatorsService;

import java.util.Objects;

import static uk.gov.di.ipv.core.library.helpers.ApiGatewayProxyEventHelper.generateAPIGatewayProxyResponseEvent;
import static uk.gov.di.ipv.core.library.helpers.ApiGatewayProxyEventHelper.getNonRequiredHeaderByKey;

public class PutContraIndicatorsHandler
        implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {

    private static final Logger LOGGER = LogManager.getLogger();
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    private static final String GOVUK_SIGNIN_JOURNEY_ID_HEADER = "govuk-signin-journey-id";
    private static final String IP_ADDRESS_HEADER = "ip-address";
    private static final String SUCCESS_RESULT = "success";
    private static final String FAIL_RESULT = "fail";

    private final ContraIndicatorsService cimitService;

    public PutContraIndicatorsHandler() {
        this.cimitService = new ContraIndicatorsService();
    }

    public PutContraIndicatorsHandler(ContraIndicatorsService cimitService) {
        this.cimitService = cimitService;
    }

    @Override
    public APIGatewayProxyResponseEvent handleRequest(
            APIGatewayProxyRequestEvent input, Context context) {
        LOGGER.info(new StringMapMessage().with("Function invoked:", "PutContraIndicators"));

        try {
            var parsedRequest = getParsedRequest(input);
            cimitService.addUserCis(parsedRequest);

            return generateAPIGatewayProxyResponseEvent(
                    200,
                    PutContraIndicatorsResponse.builder()
                            .result(SUCCESS_RESULT)
                            .reason(null)
                            .errorMessage(null)
                            .build());
        } catch (FailedToParseRequestException e) {
            LOGGER.error(
                    new StringMapMessage().with("Unable to parse input request", e.getMessage()));
            return generateAPIGatewayProxyResponseEvent(
                    400,
                    PutContraIndicatorsResponse.builder()
                            .result(FAIL_RESULT)
                            .reason(e.getClass().getSimpleName())
                            .errorMessage(e.getMessage())
                            .build());
        } catch (CiPutException e) {
            LOGGER.error(
                    new StringMapMessage()
                            .with(
                                    "CI Codes could not be inserted into the Cimit Stub Table.",
                                    e.getMessage()));

            return generateAPIGatewayProxyResponseEvent(
                    500,
                    PutContraIndicatorsResponse.builder()
                            .result(FAIL_RESULT)
                            .reason(e.getClass().getSimpleName())
                            .errorMessage(e.getMessage())
                            .build());
        } catch (Exception e) {
            LOGGER.error(
                    new StringMapMessage()
                            .with("Failed to insert CI due to unexpected error.", e.getMessage()));

            return generateAPIGatewayProxyResponseEvent(
                    500,
                    PutContraIndicatorsResponse.builder()
                            .result(FAIL_RESULT)
                            .reason(e.getClass().getSimpleName())
                            .errorMessage(e.getMessage())
                            .build());
        }
    }

    private PutContraIndicatorsRequest getParsedRequest(APIGatewayProxyRequestEvent input)
            throws FailedToParseRequestException {
        try {
            var requestBody = input.getBody();

            if (Objects.isNull(requestBody)) {
                throw new FailedToParseRequestException("Missing request body");
            }

            var parsedBody =
                    OBJECT_MAPPER.readValue(requestBody, PutContraIndicatorsRequestBody.class);
            var signedJwt = parsedBody.getSignedJwt();

            if (StringUtils.isBlank(signedJwt)) {
                throw new FailedToParseRequestException("signed_jwt is empty");
            }

            return PutContraIndicatorsRequest.builder()
                    .govukSigninJourneyId(
                            getNonRequiredHeaderByKey(GOVUK_SIGNIN_JOURNEY_ID_HEADER, input))
                    .ipAddress(getNonRequiredHeaderByKey(IP_ADDRESS_HEADER, input))
                    .signedJwt(signedJwt)
                    .build();

        } catch (JsonProcessingException e) {
            throw new FailedToParseRequestException(
                    String.format("Failed to parse request body: %s", e.getMessage()));
        }
    }
}

File: di-ipv-cimit-stub-ts/lambdas/src/internal-api/put-contra-indicators/putContraIndicatorsHandler.ts
import { APIGatewayProxyEvent, APIGatewayProxyResultV2, Context } from "aws-lambda";
import { buildApiResponse, getErrorMessage } from "../../common/apiResponseBuilder";
import { FailedToParseRequestError } from "../../common/exceptions";
import { addUserCIs } from "../../common/contraIndicatorsService";


const FAILURE_RESPONSE = "fail";
const SUCCESS_RESPONSE = "success";


export const putContraIndicatorsHandler = async (
  request: APIGatewayProxyEvent, context: Context
): Promise<APIGatewayProxyResultV2> => {
    try {
        const parsedRequest = parseRequest(request);
        addUserCIs(parsedRequest);
        return buildApiResponse(200, {result: SUCCESS_RESPONSE});
    }
    catch (error) {
      if (error instanceof FailedToParseRequestError) {
        console.error(getErrorMessage(error));
        return buildApiResponse(400, {
            result: FAILURE_RESPONSE,
            errorMessage: getErrorMessage(error),
        });
      } else {
        console.error(getErrorMessage(error));
        return buildApiResponse(500, {
            result: FAILURE_RESPONSE,
            errorMessage: getErrorMessage(error),
        });
      }
    }

}

const parseRequest = (
  request: APIGatewayProxyEvent,
): PutContraIndicatorRequest => {
  const requestBody = request.body;
  if (!requestBody) {
    console.error("Missing request body");
    throw new FailedToParseRequestError("Missing request body");
  }
  const signedJwt = JSON.parse(requestBody).signed_jwt;
  if (!signedJwt) {
    console.error("signed_jwt is empty");
    throw new FailedToParseRequestError("signed_jwt is empty");
  }
  const headers = request.headers;
  return {
    govuk_signin_journey_id: headers["govuk_signin_journey_id"],
    ip_address: headers["ip-address"],
    signed_jwt: signedJwt,
  };
}

export interface PutContraIndicatorRequest {
  govuk_signin_journey_id?: string;
  ip_address?: string
  signed_jwt: string;
}

